crates/bevy_pbr/src/material.rs

```diff
@@ -361,7 +371,14 @@ pub fn queue_material_meshes<M: SpecializedMaterial>(
 
         let inverse_view_matrix = view.transform.compute_matrix().inverse();
         let inverse_view_row_2 = inverse_view_matrix.row(2);
-        let msaa_key = MeshPipelineKey::from_msaa_samples(msaa.samples);
+
+        let mut view_key =
+            MeshPipelineKey::from_msaa_samples(msaa.samples) | MeshPipelineKey::from_hdr(view.hdr);
+        if let Some(tonemapping) = tonemapping {
+            if tonemapping.is_enabled && !view.hdr {
```

Comment:
I think it would make sense to do the tonemapping in the main pass. It can then be used in the proper location between other post process things like SMAA/FXAA, bloom, dithering, etc...

Reply:
Did you mean to say _not_ do tonemapping in the main pass?

Reply:
I think so, now thinking about it. Is the main pass the pass in the material shader? I mean doing the tone mapping as part of post process, rather than in the material. An example would be:
`material -> srgb (32bit/16bit) -> (post process) -> fxaa/smaa -> linear -> bloom -> tonemap -> srgb -> dither (8bit)`

Reply:
The idea with going to srgb is to deal with the MSAA not working well on HDR data. I don't really know if just srgb is the best option though.

Reply:
HDR/LDR and non-linear sRGB are separate things. We do lighting using physical units which means full dynamic range. Tonemapping compresses the dynamic range to fit into 0-1. sRGB requires gamma correction which makes things a bit dimmer I think, but it’s the same dynamic range.

Reply:
> The idea with going to srgb is to deal with the MSAA not working well on HDR data. I don't really know if just srgb is the best option though.

Also on that note, that means the MSAA resolve should come after tonemapping. But there was discussion earlier in the PR about how to handle this - basically doing a different and reversible tonemapping to allow MSAA resolution, and then doing the inverse of that tonemapping on the resolved texture. See https://github.com/bevyengine/bevy/pull/3425#issuecomment-1155790933 for more.

Reply:
I think this makes sense. Also the tonemapping done at this stage doesn't need to be the same as the tonemapping done in the end. It could be one thing to make MSAA work well, reversed in post processing, then something more elaborate if the user wants to change the post process. I could be wrong but I think we'd want to use Rgba16Float so that it can be more accurately reversed for things like bloom where you want to be able have the ability to only bloom on extremely bright things.

Reply:
It looks like AMD (gpuopen) has a good article on optimized reversible tonemapping that specifically talks about the MSAA issue: https://gpuopen.com/learn/optimized-reversible-tonemapper-for-resolve/

Reply:
Oh nice! @cart check out this alternative from Timothy Lottes at AMD.

Reply:
Ooh good to be aware of!

Reply:
> It may sound stupid but is there any argument for HDR (which means an Rgba16Float texture) with tonemapping in the main pass to avoid a separate pass but retain higher precision?

Probably not stupid. It just increases the complexity of the system / the complexity of extending it. Honestly having two different tonemapping / hdr paradigms in two different places (especially in shaders ... which might be user defined) is already a bit too much complexity. Introducing another fork in the config (hdr can have in-shader tonemapping) will probably make this even more of a nightmare.

I think in my ideal world, we learn that we can safely (and performantly) use one implementation (with the same assumptions) everywhere. If that doesn't pan out, ideally we build abstractions in a way that all pipeline extensions still work the same across all hdr/tonemapping configurations.